File: src/agy_swap/store.py
"""Account store CRUD, JSON schema validation, and legacy quota migration."""

from datetime import datetime, timezone
import json
import os
import re

import agy_swap
from agy_swap import ACCOUNTS_FILE, QUOTA_SCHEMA, MAX_LIMIT_SECS, Accounts, AccountStoreError
from agy_swap.display import clean_display_text, normalize_email, _parse_utc_datetime
from agy_swap.logs import auto_scan_logs_for_limits, _model_identity
from agy_swap.oauth import decode_token, extract_verified_google_email_claim
from agy_swap.storage import _atomic_write_bytes, _accounts_lock


def _normalize_quota_snapshot(snapshot, email):
    if not isinstance(snapshot, dict):
        raise AccountStoreError(f"Invalid quota snapshot for {email}")
    try:
        observed_at = _parse_utc_datetime(snapshot["observed_at"]).isoformat()
    except (KeyError, AttributeError, TypeError, ValueError):
        raise AccountStoreError(f"Invalid quota snapshot timestamp for {email}") from None

    tier = snapshot.get("tier")
    if not isinstance(tier, dict):
        raise AccountStoreError(f"Invalid quota tier for {email}")
    tier_id = clean_display_text(tier.get("id"))
    tier_name = clean_display_text(tier.get("name"))
    if not tier_id or not tier_name:
        raise AccountStoreError(f"Invalid quota tier for {email}")

    raw_groups = snapshot.get("groups")
    if not isinstance(raw_groups, list) or not raw_groups:
        raise AccountStoreError(f"Invalid quota groups for {email}")
    groups = []
    for group in raw_groups[:8]:
        if not isinstance(group, dict) or group.get("id") not in ("gemini", "third_party"):
            raise AccountStoreError(f"Invalid quota group for {email}")
        name = clean_display_text(group.get("name"))
        raw_buckets = group.get("buckets")
        if not name or not isinstance(raw_buckets, list) or not raw_buckets:
            raise AccountStoreError(f"Invalid quota group for {email}")
        buckets = []
        for bucket in raw_buckets[:8]:
            if not isinstance(bucket, dict) or bucket.get("window") not in ("weekly", "5h"):
                raise AccountStoreError(f"Invalid quota bucket for {email}")
            bucket_id = clean_display_text(bucket.get("id"))
            bucket_name = clean_display_text(bucket.get("name"))
            fraction = bucket.get("remaining_fraction")
            if not bucket_id or not bucket_name or isinstance(fraction, bool) or not isinstance(fraction, (int, float)) or not 0 <= fraction <= 1:
                raise AccountStoreError(f"Invalid quota bucket for {email}")
            try:
                reset_at = _parse_utc_datetime(bucket["reset_at"]).isoformat()
            except (KeyError, AttributeError, TypeError, ValueError):
                raise AccountStoreError(f"Invalid quota reset for {email}") from None
            buckets.append({
                "id": bucket_id,
                "name": bucket_name,
                "window": bucket["window"],
                "remaining_fraction": float(fraction),
                "reset_at": reset_at,
            })
        groups.append({"id": group["id"], "name": name, "buckets": buckets})

    return {
        "observed_at": observed_at,
        "tier": {"id": tier_id, "name": tier_name},
        "groups": groups,
    }


def _validate_accounts(data):
    if not isinstance(data, dict):
        raise AccountStoreError("accounts.json must contain an object")
    accounts = Accounts()
    for key, value in data.items():
        if not isinstance(key, str) or not isinstance(value, dict):
            raise AccountStoreError("accounts.json contains an invalid account entry")
        email = normalize_email(value.get("email") or key)
        if not email:
            raise AccountStoreError(f"Invalid account email: {key!r}")
        if email in accounts:
            raise AccountStoreError(f"Duplicate account email: {email}")
        account = dict(value)
        account["email"] = email
        account["name"] = clean_display_text(account.get("name"), "Google User") or "Google User"
        token = account.get("token_data")
        if token is not None and (not isinstance(token, str) or not decode_token(token)):
            raise AccountStoreError(f"Invalid saved token for {email}")
        claimed_email = extract_verified_google_email_claim(token)
        if claimed_email and claimed_email != email:
            raise AccountStoreError(f"Saved token email does not match {email}")
        quota_limits = account.get("quota_limits", {})
        if not isinstance(quota_limits, dict):
            raise AccountStoreError(f"Invalid quota limits for {email}")
        normalized_limits = {}
        for key, limit in quota_limits.items():
            if not isinstance(key, str) or not isinstance(limit, dict):
                raise AccountStoreError(f"Invalid quota limit for {email}")
            model = clean_display_text(limit.get("model"))
            family = limit.get("family")
            source = limit.get("source")
            if not model or family not in ("claude", "gemini", "gpt") or source not in ("log", "manual"):
                raise AccountStoreError(f"Invalid quota metadata for {email}")
            try:
                reset_at = _parse_utc_datetime(limit.get("reset_at")).isoformat()
                observed_at = _parse_utc_datetime(limit.get("observed_at")).isoformat()
            except (AttributeError, TypeError, ValueError):
                raise AccountStoreError(f"Invalid quota timestamp for {email}") from None
            normalized_limits[key] = {
                "model": model,
                "family": family,
                "reset_at": reset_at,
                "observed_at": observed_at,
                "source": source,
            }
            source_file = clean_display_text(limit.get("source_file"))
            if source_file:
                normalized_limits[key]["source_file"] = re.split(r"[\\/]", source_file)[-1]
        if normalized_limits:
            account["quota_limits"] = normalized_limits
        else:
            account.pop("quota_limits", None)
        if "legacy_quota" in account and not isinstance(account["legacy_quota"], dict):
            raise AccountStoreError(f"Invalid legacy quota for {email}")
        if "quota_snapshot" in account:
            account["quota_snapshot"] = _normalize_quota_snapshot(account["quota_snapshot"], email)
        accounts[email] = account
    return accounts


def migrate_legacy_quota(account):
    if account.get("quota_schema") == QUOTA_SCHEMA:
        return False
    legacy_keys = (
        "limit_reset", "limit_reset_claude", "limit_reset_gemini",
        "claude_pct", "gemini_pct",
    )
    legacy = dict(account.get("legacy_quota", {}))
    changed = False
    for key in legacy_keys:
        if key in account:
            legacy[key] = account.pop(key)
            changed = True
    if legacy:
        account["legacy_quota"] = legacy
    account["quota_schema"] = QUOTA_SCHEMA
    return True


def reconcile_log_limits(account, detected, evidence):
    limits = account.get("quota_limits", {})
    changed = False
    for key, limit in list(limits.items()):
        if limit.get("source") != "log":
            continue
        event = evidence.get(_model_identity(limit.get("model", "")))
        if not event:
            continue
        if _parse_utc_datetime(event["observed_at"]) >= _parse_utc_datetime(limit["observed_at"]):
            del limits[key]
            changed = True

    for key, limit in detected.items():
        existing = limits.get(key)
        if not existing or _parse_utc_datetime(limit["observed_at"]) > _parse_utc_datetime(existing["observed_at"]):
            limits[key] = limit
            changed = True

    if limits:
        account["quota_limits"] = limits
    else:
        account.pop("quota_limits", None)
    return changed


def load_accounts(sync_logs=True):
    try:
        revision = os.stat(ACCOUNTS_FILE).st_mtime_ns
        with open(ACCOUNTS_FILE, "r", encoding="utf-8") as f:
            accounts = _validate_accounts(json.load(f))
    except FileNotFoundError:
        revision = None
        accounts = Accounts()
    except (OSError, UnicodeDecodeError, json.JSONDecodeError, AccountStoreError) as exc:
        raise AccountStoreError(f"Cannot read {ACCOUNTS_FILE}: {exc}") from exc
    accounts.revision = revision

    if sync_logs and accounts:
        changed = False
        for acc in accounts.values():
            changed = migrate_legacy_quota(acc) or changed
        now = datetime.now(timezone.utc)

        for acc in accounts.values():
            limits = acc.get("quota_limits", {})
            for key, limit in list(limits.items()):
                dt = _parse_utc_datetime(limit["reset_at"])
                if dt <= now or (dt - now).total_seconds() > MAX_LIMIT_SECS:
                    del limits[key]
                    changed = True
            if not limits:
                acc.pop("quota_limits", None)

        detected_limits, evidence = auto_scan_logs_for_limits(include_evidence=True)
        for email, acc in accounts.items():
            changed = reconcile_log_limits(acc, detected_limits.get(email, {}), evidence.get(email, {})) or changed

        if changed:
            save_accounts(accounts)

    return accounts


def save_accounts(accounts):
    validated = _validate_accounts(accounts)
    payload = (json.dumps(validated, indent=2, ensure_ascii=False) + "\n").encode("utf-8")
    with _accounts_lock():
        try:
            current_revision = os.stat(ACCOUNTS_FILE).st_mtime_ns
        except FileNotFoundError:
            current_revision = None
        expected_revision = getattr(accounts, "revision", current_revision)
        if expected_revision != current_revision:
            raise AccountStoreError("accounts.json changed in another process; retry the command")
        if current_revision is not None:
            with open(ACCOUNTS_FILE, "rb") as f:
                _atomic_write_bytes(ACCOUNTS_FILE + ".bak", f.read())
        _atomic_write_bytes(ACCOUNTS_FILE, payload)
        accounts.revision = os.stat(ACCOUNTS_FILE).st_mtime_ns

File: internal/app/store.go
package app

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type Store struct {
	paths   Paths
	scanner *LogScanner
}

func NewStore(paths Paths) *Store {
	store := &Store{paths: paths}
	store.scanner = NewLogScanner(paths)
	return store
}

func (s *Store) Load(syncLogs bool) (*Accounts, error) {
	data, err := os.ReadFile(s.paths.Accounts)
	if errors.Is(err, os.ErrNotExist) {
		return NewAccounts(), nil
	}
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %w", s.paths.Accounts, err)
	}
	accounts, err := decodeOrderedAccounts(data)
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %w", s.paths.Accounts, err)
	}
	if stat, statErr := os.Stat(s.paths.Accounts); statErr == nil {
		accounts.Revision = stat.ModTime().UnixNano()
	}
	if syncLogs && accounts.Len() > 0 {
		changed := migrateAndExpire(accounts, time.Now().UTC())
		limits, evidence, scanErr := s.scanner.Scan()
		if scanErr == nil {
			for _, email := range accounts.Order {
				account := accounts.ByEmail[email]
				if reconcileLogLimits(account, limits[email], evidence[email]) {
					changed = true
				}
			}
		}
		if changed {
			if saveErr := s.Save(accounts); saveErr != nil && !errors.Is(saveErr, errStoreConflict) {
				return nil, saveErr
			}
		}
	}
	return accounts, nil
}

func (s *Store) Save(accounts *Accounts) error {
	if err := validateAccounts(accounts); err != nil {
		return err
	}
	payload, err := encodeOrderedAccounts(accounts)
	if err != nil {
		return err
	}
	lock, err := acquireFileLock(s.paths.AccountsLock)
	if err != nil {
		return err
	}
	defer lock.Close()
	currentRevision := int64(0)
	if stat, statErr := os.Stat(s.paths.Accounts); statErr == nil {
		currentRevision = stat.ModTime().UnixNano()
	} else if !errors.Is(statErr, os.ErrNotExist) {
		return statErr
	}
	if accounts.Revision != 0 && accounts.Revision != currentRevision {
		return errStoreConflict
	}
	if currentRevision != 0 {
		previous, readErr := os.ReadFile(s.paths.Accounts)
		if readErr != nil {
			return readErr
		}
		if err := atomicWrite(s.paths.AccountsBackup, previous, 0o600); err != nil {
			return err
		}
	}
	if err := atomicWrite(s.paths.Accounts, payload, 0o600); err != nil {
		return err
	}
	stat, err := os.Stat(s.paths.Accounts)
	if err != nil {
		return err
	}
	accounts.Revision = stat.ModTime().UnixNano()
	return nil
}

func validateAccounts(accounts *Accounts) error {
	seen := make(map[string]struct{}, accounts.Len())
	for _, key := range accounts.Order {
		account, ok := accounts.ByEmail[key]
		if !ok || account == nil {
			return fmt.Errorf("accounts.json contains an invalid account entry")
		}
		email := normalizeEmail(firstString(account["email"], key))
		if email == "" {
			return fmt.Errorf("invalid account email: %q", key)
		}
		if _, exists := seen[email]; exists {
			return fmt.Errorf("duplicate account email: %s", email)
		}
		seen[email] = struct{}{}
		account["email"] = email
		account["name"] = firstString(account["name"], "Google User")
		if token, ok := account["token_data"].(string); ok {
			if decodeToken(token) == nil {
				return fmt.Errorf("invalid saved token for %s", email)
			}
			if claimed := extractVerifiedEmail(token); claimed != "" && claimed != email {
				return fmt.Errorf("saved token email does not match %s", email)
			}
		}
		if err := normalizeQuotaLimits(account, email); err != nil {
			return err
		}
		if raw, ok := account["quota_snapshot"]; ok {
			norm, err := normalizeQuotaSnapshot(raw, email)
			if err != nil {
				return err
			}
			account["quota_snapshot"] = norm
		}
	}
	return nil
}

func normalizeQuotaLimits(account Account, email string) error {
	raw, exists := account["quota_limits"]
	if !exists || raw == nil {
		return nil
	}
	limits, ok := raw.(map[string]any)
	if !ok {
		return fmt.Errorf("invalid quota limits for %s", email)
	}
	normalized := make(map[string]any)
	for key, rawLimit := range limits {
		limit, ok := rawLimit.(map[string]any)
		if !ok {
			return fmt.Errorf("invalid quota limit for %s", email)
		}
		model := cleanText(getString(limit, "model"))
		family := getString(limit, "family")
		source := getString(limit, "source")
		if model == "" || !oneOf(family, "claude", "gemini", "gpt") || !oneOf(source, "log", "manual") {
			return fmt.Errorf("invalid quota metadata for %s", email)
		}
		resetAt, err1 := parseUTC(getString(limit, "reset_at"))
		observedAt, err2 := parseUTC(getString(limit, "observed_at"))
		if err1 != nil || err2 != nil {
			return fmt.Errorf("invalid quota timestamp for %s", email)
		}
		norm := map[string]any{
			"model": model, "family": family, "reset_at": isoTime(resetAt),
			"observed_at": isoTime(observedAt), "source": source,
		}
		if sourceFile := filepath.Base(strings.ReplaceAll(cleanText(getString(limit, "source_file")), "\\", "/")); sourceFile != "." && sourceFile != "" {
			norm["source_file"] = sourceFile
		}
		normalized[key] = norm
	}
	if len(normalized) == 0 {
		delete(account, "quota_limits")
	} else {
		account["quota_limits"] = normalized
	}
	return nil
}

func normalizeQuotaSnapshot(raw any, email string) (map[string]any, error) {
	snapshot, ok := raw.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("invalid quota snapshot for %s", email)
	}
	observed, err := parseUTC(getString(snapshot, "observed_at"))
	if err != nil {
		return nil, fmt.Errorf("invalid quota snapshot timestamp for %s", email)
	}
	tier := getMap(snapshot["tier"])
	tierID, tierName := cleanText(getString(tier, "id")), cleanText(getString(tier, "name"))
	if tierID == "" || tierName == "" {
		return nil, fmt.Errorf("invalid quota tier for %s", email)
	}
	rawGroups := getSlice(snapshot["groups"])
	if len(rawGroups) == 0 {
		return nil, fmt.Errorf("invalid quota groups for %s", email)
	}
	groups := make([]any, 0, minInt(len(rawGroups), 8))
	for _, rawGroup := range rawGroups[:minInt(len(rawGroups), 8)] {
		group := getMap(rawGroup)
		groupID, name := getString(group, "id"), cleanText(getString(group, "name"))
		bucketsRaw := getSlice(group["buckets"])
		if !oneOf(groupID, "gemini", "third_party") || name == "" || len(bucketsRaw) == 0 {
			return nil, fmt.Errorf("invalid quota group for %s", email)
		}
		buckets := make([]any, 0, minInt(len(bucketsRaw), 8))
		for _, rawBucket := range bucketsRaw[:minInt(len(bucketsRaw), 8)] {
			bucket := getMap(rawBucket)
			window := getString(bucket, "window")
			id, bucketName := cleanText(getString(bucket, "id")), cleanText(getString(bucket, "name"))
			fraction, validFraction := getFloat(bucket["remaining_fraction"])
			reset, resetErr := parseUTC(getString(bucket, "reset_at"))
			if !oneOf(window, "weekly", "5h") || id == "" || bucketName == "" || !validFraction || fraction < 0 || fraction > 1 || resetErr != nil {
				return nil, fmt.Errorf("invalid quota bucket for %s", email)
			}
			buckets = append(buckets, map[string]any{"id": id, "name": bucketName, "window": window, "remaining_fraction": fraction, "reset_at": isoTime(reset)})
		}
		groups = append(groups, map[string]any{"id": groupID, "name": name, "buckets": buckets})
	}
	return map[string]any{"observed_at": isoTime(observed), "tier": map[string]any{"id": tierID, "name": tierName}, "groups": groups}, nil
}

func migrateAndExpire(accounts *Accounts, now time.Time) bool {
	changed := false
	legacyKeys := []string{"limit_reset", "limit_reset_claude", "limit_reset_gemini", "claude_pct", "gemini_pct"}
	for _, email := range accounts.Order {
		account := accounts.ByEmail[email]
		schema, _ := numberInt(account["quota_schema"])
		if schema != quotaSchema {
			legacy := getMap(account["legacy_quota"])
			if legacy == nil {
				legacy = make(map[string]any)
			}
			for _, key := range legacyKeys {
				if value, ok := account[key]; ok {
					legacy[key] = value
					delete(account, key)
					changed = true
				}
			}
			if len(legacy) > 0 {
				account["legacy_quota"] = legacy
			}
			account["quota_schema"] = quotaSchema
			changed = true
		}
		limits := getMap(account["quota_limits"])
		for key, raw := range limits {
			limit := getMap(raw)
			reset, err := parseUTC(getString(limit, "reset_at"))
			if err != nil || !reset.After(now) || reset.Sub(now) > maxLimitDuration {
				delete(limits, key)
				changed = true
			}
		}
		if len(limits) == 0 {
			delete(account, "quota_limits")
		}
	}
	return changed
}

func reconcileLogLimits(account Account, detected map[string]LimitRecord, evidence map[string]EvidenceRecord) bool {
	limits := getMap(account["quota_limits"])
	if limits == nil {
		limits = make(map[string]any)
	}
	changed := false
	for key, raw := range limits {
		limit := getMap(raw)
		if getString(limit, "source") != "log" {
			continue
		}
		event, ok := evidence[modelIdentity(getString(limit, "model"))]
		if !ok {
			continue
		}
		eventTime, _ := parseUTC(event.ObservedAt)
		limitTime, _ := parseUTC(getString(limit, "observed_at"))
		if !eventTime.Before(limitTime) {
			delete(limits, key)
			changed = true
		}
	}
	for key, record := range detected {
		existing := getMap(limits[key])
		newTime, _ := parseUTC(record.ObservedAt)
		oldTime, oldErr := parseUTC(getString(existing, "observed_at"))
		if existing == nil || oldErr != nil || newTime.After(oldTime) {
			limits[key] = record.Map()
			changed = true
		}
	}
	if len(limits) == 0 {
		delete(account, "quota_limits")
	} else {
		account["quota_limits"] = limits
	}
	return changed
}

func oneOf(value string, options ...string) bool {
	for _, option := range options {
		if value == option {
			return true
		}
	}
	return false
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
